loki/bulk/scheduler.py

```diff
@@ -407,6 +407,7 @@ def process(self, transformation, reverse=False):
         always processed before their target :any:`Subroutine`.
         """
 
+        root = list(nx.topological_sort(self.item_graph))[0]
```

Comment:
Yes. For the current implementation I don't need the depth. However, for the CUF transformation I will need the depth to know whether a function is a kernel (`__global__` subroutine) or a device function (a function called within a kernel function, thus a `__device__` subroutine). 

Having one than more root a possible implementation would be/include (e.g. within the `Scheduler` class `__init__` method):

```py
topological_generations = list(nx.topological_generations(self.item_graph))
# all of the generations/depths, this gives the maximum depth per item if the subroutine has multiple depths
for i_gen, gen in enumerate(topological_generations):
    print("{}: {}".format(i_gen, gen))
# those are all roots
roots = topological_generations[0]
# in order to get all the depths per item (and per root)
depths = {}
for item in self.item_graph:
    depths[item] = {}
    depths[item]["all"] = []
    for root in topological_generations[0]:
        simple_paths = list(nx.all_simple_paths(self.item_graph, source=root, target=item))
        depths[item]["all"].extend([len(_path) for _path in simple_paths])
        depths[item]["root"] = {}
        depths[item]["root"][root] = [len(_path) for _path in simple_paths]
```

For my purposes it will be sufficient to just know whether a subroutine is called by a driver (e.g. `level = 1`) or called by a kernel (e.g. `level >= 2`). 
**However, introducing this depth information for any transformation we may consider to pass/collect more information than that?** I would like to discuss which information and how to store/pass those information.
Nonetheless, I would suppose to exclude this from this pull request and make another pull request for this or include it in the CUF transformation pull request.

Reply:
Agreed, the question of what information to collect and pass mandates a more in-depth discussion.

I'm fairly far out of my area of expertise with graph algorithms, but I wonder if it would make sense to maintain also an adjoint view of the item_graph? That way, `all_simple_paths` might be able to produce the information for all roots in a single call (just viewed in the reverse direction).